### ISSUE TITLE
Is projector also used for ibot?

### ISSUE THREAD

**Issue by qimaqi** [RobinWu218/SimDINO#4]
In the paper, under Model Architecture in Section 3.1 Experimental Setup, it mentions that an MLP with a hidden size of 2048 and an output dimension of 256 is used as a projector. Could you explain this further? Is this projector also used for patch feature to calculate ibot loss? It is not clearly stated in the appendix pseudocode. Did L2 normalization also used for patch feature or not? 
Nice work!

**Comment by RobinWu218** [comment]
Yes, the projector g is connected to the ViT backbone f and is used to produce feature z in the pseudocode. To make it more explicit, here is how we go from input x to feature z: x -> f -> g -> L2-norm -> z. All features are produced this way, including patch features for ibot. Hope this clarification helps.

**Comment by qimaqi** [comment]
Thanks for clarification:)
 I would also like to ask if a shared projector is used for patch feature and the global feature(from cls token). In the appendix it mentions iBOT head tying is False so I guess a shared project (3 layer MLP) is used for both? I am wondering if it would be possible to provide the training logs later, as I noticed that the L2 loss of iBOT is very small.

**Comment by RobinWu218** [comment]
That is true. We use a shared projector for both iBOT and DINO in our experiments, although it can be decoupled. I am not sure what you mean by "L2 loss of iBOT is small": we simply calculate the average loss per patch token, the same way we compute the DINO loss (per cls token).

**Comment by qimaqi** [comment]
Oh, so for each patch, there is also a coding rate loss in the iBOT part, as shown in Equation 8?
For the L2 loss, I am referring to Equation 7
So, it treats each patch token like a CLS token, calculates the loss in the same way like equation 7 and 8 shows, and then averages?
Thanks for reply

**Comment by RobinWu218** [comment]
Sorry for my late reply. The coding rate is **not** computed for each patch. Please refer to Eq. 17, where we define the entire loss for SimDINOv2. Note we compute the coding rate on the cls token on the global views. For iBOT, we average the L2 loss between the (masked) patch features of student and the unmasked patch features of teacher. Hope the clarification helps. Meanwhile, we will soon finish cleaning up the codebase and release it. Thanks for your patience.

**Comment by qimaqi** [comment]
It is clear now with code, thanks!

**Comment by Aceticia** [comment]
The code is pretty clean and thanks again for the nice repo. I wonder why you chose to project to 2048 and down to 256, rather than using the original size right out of the model? Doesn't this introduce more hyperparameters?

**Comment by RobinWu218** [comment]
This structure is called a projector and is quite common in SSL methods. For fair comparison, here we simply adopt the same projector design as in the original [DINO](https://github.com/facebookresearch/dino/blob/7c446df5b9f45747937fb0d72314eb9f7b66930a/vision_transformer.py#L261-L274). For a more thorough study on the role of projectors, here is a nice [paper](https://openreview.net/pdf?id=kDEL91Dufpa)

**Comment by qimaqi** [comment]
For projector one more question is about the option: remove_last_layer 
I see by default it is true, so the output dimension is not compressed but to higher dimension, is it correct?

**Comment by Aceticia** [comment]
I believe what's removed is the "d->number of prototypes" part of the original DINO head

**Comment by RobinWu218** [comment]
@Aceticia is correct. By setting remove_last_layer=True, we remove the last layer that projects to high dimension (in DINO).
